// @nestjs/graphql/dist/services/resolvers-explorer.service.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const common_1 = require("@nestjs/common");
const shared_utils_1 = require("@nestjs/common/utils/shared.utils");
const core_1 = require("@nestjs/core");
const context_id_factory_1 = require("@nestjs/core/helpers/context-id-factory");
const external_context_creator_1 = require("@nestjs/core/helpers/external-context-creator");
const injector_1 = require("@nestjs/core/injector/injector");
const internal_core_module_1 = require("@nestjs/core/injector/internal-core-module");
const modules_container_1 = require("@nestjs/core/injector/modules-container");
const metadata_scanner_1 = require("@nestjs/core/metadata-scanner");
const apollo_server_express_1 = require("apollo-server-express");
const lodash_1 = require("lodash");
const gql_paramtype_enum_1 = require("../enums/gql-paramtype.enum");
const resolvers_enum_1 = require("../enums/resolvers.enum");
const params_factory_1 = require("../factories/params.factory");
const graphql_constants_1 = require("../graphql.constants");
const extract_metadata_util_1 = require("../utils/extract-metadata.util");
const base_explorer_service_1 = require("./base-explorer.service");
let ResolversExplorerService = class ResolversExplorerService extends base_explorer_service_1.BaseExplorerService {
    constructor(modulesContainer, metadataScanner, externalContextCreator, gqlOptions) {
        super();
        this.modulesContainer = modulesContainer;
        this.metadataScanner = metadataScanner;
        this.externalContextCreator = externalContextCreator;
        this.gqlOptions = gqlOptions;
        this.gqlParamsFactory = new params_factory_1.GqlParamsFactory();
        this.injector = new injector_1.Injector();
    }
    explore() {
        const modules = this.getModules(this.modulesContainer, this.gqlOptions.include || []);
        const resolvers = this.flatMap(modules, (instance, moduleRef) => this.filterResolvers(instance, moduleRef));
        return this.groupMetadata(resolvers);
    }
    filterResolvers(wrapper, moduleRef) {
        const { instance } = wrapper;
        if (!instance) {
            return undefined;
        }
        const prototype = Object.getPrototypeOf(instance);
        const predicate = (resolverType, isDelegated, isPropertyResolver) => shared_utils_1.isUndefined(resolverType) ||
            isDelegated ||
            (!isPropertyResolver &&
                ![resolvers_enum_1.Resolvers.MUTATION, resolvers_enum_1.Resolvers.QUERY, resolvers_enum_1.Resolvers.SUBSCRIPTION].some(type => type === resolverType));
        const resolvers = this.metadataScanner.scanFromPrototype(instance, prototype, name => extract_metadata_util_1.extractMetadata(instance, prototype, name, predicate));
        const isRequestScoped = !wrapper.isDependencyTreeStatic();
        return resolvers
            .filter(resolver => !!resolver)
            .map(resolver => {
            const createContext = (transform) => this.createContextCallback(instance, prototype, wrapper, moduleRef, resolver, isRequestScoped, transform);
            if (resolver.type === graphql_constants_1.SUBSCRIPTION_TYPE) {
                const subscriptionOptions = Reflect.getMetadata(graphql_constants_1.SUBSCRIPTION_OPTIONS_METADATA, instance[resolver.methodName]);
                return this.createSubscriptionMetadata(createContext, subscriptionOptions, resolver);
            }
            return Object.assign({}, resolver, { callback: createContext() });
        });
    }
    createContextCallback(instance, prototype, wrapper, moduleRef, resolver, isRequestScoped, transform = lodash_1.identity) {
        const paramsFactory = this.gqlParamsFactory;
        if (isRequestScoped) {
            const resolverCallback = (...args) => __awaiter(this, void 0, void 0, function* () {
                const gqlContext = paramsFactory.exchangeKeyForValue(gql_paramtype_enum_1.GqlParamtype.CONTEXT, undefined, args);
                const gqlInfo = paramsFactory.exchangeKeyForValue(gql_paramtype_enum_1.GqlParamtype.INFO, undefined, args);
                const contextId = gqlInfo && gqlInfo.contextId ? gqlInfo.contextId : context_id_factory_1.createContextId();
                gqlInfo && (gqlInfo.contextId = contextId);
                this.registerContextProvider(gqlContext, contextId);
                const contextInstance = yield this.injector.loadPerContext(instance, moduleRef, moduleRef.providers, contextId);
                const callback = this.externalContextCreator.create(contextInstance, transform(contextInstance[resolver.methodName]), resolver.methodName, graphql_constants_1.PARAM_ARGS_METADATA, paramsFactory, contextId, wrapper.id);
                return callback(...args);
            });
            return resolverCallback;
        }
        const resolverCallback = this.externalContextCreator.create(instance, prototype[resolver.methodName], resolver.methodName, graphql_constants_1.PARAM_ARGS_METADATA, paramsFactory);
        return resolverCallback;
    }
    createSubscriptionMetadata(createSubscribeContext, subscriptionOptions, resolverMetadata) {
        const baseCallbackMetadata = {
            resolve: subscriptionOptions && subscriptionOptions.resolve,
        };
        if (subscriptionOptions && subscriptionOptions.filter) {
            return Object.assign({}, resolverMetadata, { callback: Object.assign({}, baseCallbackMetadata, { subscribe: createSubscribeContext((subscription) => (...args) => apollo_server_express_1.withFilter(subscription(), subscriptionOptions.filter)) }) });
        }
        return Object.assign({}, resolverMetadata, { callback: Object.assign({}, baseCallbackMetadata, { subscribe: createSubscribeContext() }) });
    }
    registerContextProvider(request, contextId) {
        const coreModuleArray = [...this.modulesContainer.entries()]
            .filter(([key, { metatype }]) => metatype && metatype.name === internal_core_module_1.InternalCoreModule.name)
            .map(([key, value]) => value);
        const coreModuleRef = lodash_1.head(coreModuleArray);
        if (!coreModuleRef) {
            return;
        }
        const wrapper = coreModuleRef.getProviderByKey(core_1.REQUEST);
        wrapper.setInstanceByContextId(contextId, {
            instance: request,
            isResolved: true,
        });
    }
};
ResolversExplorerService = __decorate([
    common_1.Injectable(),
    __param(3, common_1.Inject(graphql_constants_1.GRAPHQL_MODULE_OPTIONS)),
    __metadata("design:paramtypes", [modules_container_1.ModulesContainer,
        metadata_scanner_1.MetadataScanner,
        external_context_creator_1.ExternalContextCreator, Object])
], ResolversExplorerService);
exports.ResolversExplorerService = ResolversExplorerService;
